source/common/stats/hystrix.cc

```diff
@@ -0,0 +1,315 @@
+#include "common/stats/hystrix.h"
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <sstream>
+
+#include "absl/strings/str_cat.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Stats {
+
+const uint64_t Hystrix::DEFAULT_NUM_OF_BUCKETS;
+const uint64_t Hystrix::ROLLING_WINDOW_IN_MS;
+const uint64_t Hystrix::PING_INTERVAL_IN_MS;
+
+// add new value to rolling window, in place of oldest one
+void Hystrix::pushNewValue(std::string key, uint64_t value) {
+  // create vector if do not exist
+  if (rolling_stats_map_.find(key) == rolling_stats_map_.end()) {
+    rolling_stats_map_[key].resize(num_of_buckets_, value);
+  } else {
+    rolling_stats_map_[key][current_index_] = value;
+  }
+}
+
+uint64_t Hystrix::getRollingValue(absl::string_view cluster_name, absl::string_view stats) {
+  std::string key;
+  key = absl::StrCat("cluster.",cluster_name,".",stats);
+  if (rolling_stats_map_.find(key) != rolling_stats_map_.end()) {
+    // if the counter was reset, the result is negative
+    // better return 0, will be back to normal once one rolling window passes
+    if (rolling_stats_map_[key][current_index_] <
+        rolling_stats_map_[key][(current_index_ + 1) % num_of_buckets_]) {
+      return 0;
+    } else {
+      return rolling_stats_map_[key][current_index_] -
+             rolling_stats_map_[key][(current_index_ + 1) % num_of_buckets_];
+    }
+  } else {
+    return 0;
+  }
+}
+
+void Hystrix::updateRollingWindowMap(Stats::Store& stats, absl::string_view cluster_name) {
+  std::string prefix;
+  prefix = absl::StrCat("cluster.",cluster_name,".");
+
+  // combining timeouts+retries - retries are counted  as separate requests
+  // (alternative: each request including the retries counted as 1)
+  std::string upstream_rq_timeout_key = absl::StrCat(prefix,"upstream_rq_timeout");
+  std::string upstream_rq_per_try_timeout_key = absl::StrCat(prefix,"upstream_rq_per_try_timeout");
+  uint64_t timeouts = stats.counter(upstream_rq_timeout_key).value() +
```

Comment:
If this is going to happen a bunch, do you want to cache the stats.counter return references and store them in the class, rather than doing the name-lookup each time?